File: cpp/linked_list.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

struct Node
{
public:
  int value;
  Node *next = nullptr;

  Node(int v) : value(v) {}
};

struct LinkedList
{
private:
  int size = 0;
  Node *head = nullptr;

public:
  void insert(int value)
  {
    Node *tmp = this->head;
    Node *node = new Node(value);

    do
    {
      if (this->head == nullptr)
      {
        this->head = node;
        break;
      }

      if (tmp->next == nullptr)
      {
        tmp->next = node;
        break;
      }

      tmp = tmp->next;
    } while (tmp != nullptr);
  }

  void insertAtBeginning(int value)
  {
    Node *node = new Node(value);

    node->next = this->head;
    this->head = node;
  }

  void pop()
  {
    Node *prev = nullptr;
    Node *curr = this->head;

    while (curr != nullptr)
    {
      if (curr->next == nullptr)
      {
        prev->next = nullptr;
        delete curr;
        break;
      }

      prev = curr;
      curr = curr->next;
    }
  }

  void show()
  {
    Node *tmp = this->head;

    while (tmp != nullptr)
    {
      cout << tmp->value << endl;
      tmp = tmp->next;
    }
  }
};

int main()
{
  LinkedList *list = new LinkedList();

  list->insert(10);
  list->insertAtBeginning(5);
  list->insert(20);
  list->insert(30);
  list->insertAtBeginning(2);
  list->insert(40);
  list->pop();
  list->show();

  return 0;
}

File: rust/src/linked_list.rs
use std::ptr;

struct Node {
    value: i64,
    next: *mut Node,
}

impl Node {
    pub fn new(value: i64) -> Self {
        Self {
            value,
            next: ptr::null_mut(),
        }
    }
}

struct LinkedList {
    size: u64,
    head: *mut Node,
}

impl LinkedList {
    pub fn new() -> Self {
        Self {
            size: 0,
            head: ptr::null_mut(),
        }
    }

    pub fn insert(&mut self, value: i64) {
        let node = Box::into_raw(Box::new(Node::new(value)));
        self.size += 1;

        if self.head.is_null() {
            self.head = node;
            return;
        }

        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                if (*tmp).next.is_null() {
                    (*tmp).next = node;
                    break;
                }

                tmp = (*tmp).next;
            }
        }
    }

    pub fn insertAtBeginning(&mut self, value: i64) {
        let node = Box::into_raw(Box::new(Node::new(value)));
        self.size += 1;

        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                if (*tmp).next.is_null() {
                    (*tmp).next = node;
                    break;
                }

                tmp = (*tmp).next;
            }
        }
    }

    pub fn pop(&mut self) {
        if self.head.is_null() {
            return;
        }

        let mut previous = self.head;
        let mut current = self.head;

        unsafe {
            while !current.is_null() {
                if (*current).next.is_null() {
                    (*previous).next = ptr::null_mut();
                    break;
                }

                previous = current;
                current = (*current).next;
            }
        }
    }

    pub fn show(&self) {
        let mut tmp = self.head;

        unsafe {
            while !tmp.is_null() {
                println!("{}", (*tmp).value);
                tmp = (*tmp).next;
            }
        }
    }
}
